tests/testthat/test-generate.R

```diff
@@ -1827,3 +1827,100 @@ test_that("can generate pi", {
       "  state[0] = M_PI;",
       "}"))
 })
+
+test_that("Array assignment with index on lhs and rhs (1)", {
+  ## See (mrc-5894)
+  dat <- odin_parse({
+    seed_age_band <- as.integer(4)
+    n <- parameter(type = "integer", constant = TRUE)
+    dim(x) <- n
+    initial(y) <- 0
+    update(y) <- sum(x)
+    x[] <- Poisson(1)
+    x[seed_age_band] <- x[i] + 1
+  })
+  dat <- generate_prepare(dat)
+  src <- generate_dust_system_update(dat)
+  src <- src[grepl("seed_age_band", src)]
+  expect_equal(src,
+    paste0("  internal.x[shared.seed_age_band - 1] = ",
+             "internal.x[shared.seed_age_band - 1] + 1;"))
```

Comment:
Please follow the pattern in the other tests:

```suggestion
  expect_equal(
    generate_dust_system_update(dat),
    c(method_args$update,
      "  for (size_t i = 1; i <= shared.dim.x.size; ++i) {",
      "    internal.x[i - 1] = monty::random::poisson<real_type>(rng_state, 1);",
      "  }",
      "  {",
      "    const size_t i = shared.seed_age_band;",
      "    internal.x[shared.seed_age_band - 1] = internal.x[i - 1] + 1;",
      "  }",
      "  state_next[0] = dust2::array::sum<real_type>(internal.x.data(), shared.dim.x);",
      "}"))
```

with the change above; please apply the same idea below